### ISSUE TITLE
Doxygen crash with NUM_PROC_THREADS>1

### ISSUE THREAD

**Issue by ferdymercury** [doxygen/doxygen#9155]
**Describe the bug**
When building the ROOT project documentation with doxygen, close to the end of the process, we get a crash if we set the NUM_PROC_THREADS to 7. All works perfectly fine if this flag is left at 1.

**Expected behavior**
No crash is found.

**Screenshots**
```
Generating code for file core/base/inc/RQ_OBJECT.h...
Generating code for file core/base/inc/RStipples.h...
Generating code for file core/base/inc/Rtypes.h...
Generating code for file core/base/inc/RtypesImp.h...
Generating code for file core/base/inc/Strlen.h...
Generating code for file core/base/inc/TApplication.h...
Generating code for file core/base/inc/TAtt3D.h...
Generating code for file core/base/inc/TAttAxis.h...
error: Could not open file /home/ferhue/rootdoc/html/RDrawableRequest_8cxx_source.html for writing
Exiting...
Generating code for file core/base/inc/TAttBBox.h...
error: could not open file /opt/root_src/core/base/inc/TAttBBox.h
error: cannot open file '/opt/root_src/core/base/inc/TAttBBox.h' for reading
Generating code for file core/base/inc/TAttBBox2D.h...
error: could not open file /opt/root_src/core/base/inc/TAttBBox2D.h
error: cannot open file '/opt/root_src/core/base/inc/TAttBBox2D.h' for reading
Generating code for file core/base/inc/TAttFill.h...
error: could not open file /opt/root_src/core/base/inc/TAttFill.h
error: cannot open file '/opt/root_src/core/base/inc/TAttFill.h' for reading
Generating code for file core/base/inc/TAttLine.h...
error: could not open file /opt/root_src/core/base/inc/TAttLine.h
error: cannot open file '/opt/root_src/core/base/inc/TAttLine.h' for reading
Generating code for file core/base/inc/TAttMarker.h...
error: could not open file /opt/root_src/core/base/inc/TAttMarker.h
error: cannot open file '/opt/root_src/core/base/inc/TAttMarker.h' for reading
Generating code for file core/base/inc/TAttPad.h...
error: could not open file /opt/root_src/core/base/inc/TAttPad.h
error: cannot open file '/opt/root_src/core/base/inc/TAttPad.h' for reading
Generating code for file core/base/inc/TAttText.h...
error: could not open file /opt/root_src/core/base/inc/TAttText.h
error: cannot open file '/opt/root_src/core/base/inc/TAttText.h' for reading
Generating code for file core/base/inc/TBase64.h...
error: could not open file /opt/root_src/core/base/inc/TBase64.h
error: cannot open file '/opt/root_src/core/base/inc/TBase64.h' for reading
Generating code for file core/base/inc/TBenchmark.h...
error: could not open file /opt/root_src/core/base/inc/TBenchmark.h
error: cannot open file '/opt/root_src/core/base/inc/TBenchmark.h' for reading
Generating code for file core/base/inc/TBuffer.h...
error: could not open file /opt/root_src/core/base/inc/TBuffer.h
error: cannot open file '/opt/root_src/core/base/inc/TBuffer.h' for reading
Generating code for file core/base/inc/TBuffer3D.h...
error: could not open file /opt/root_src/core/base/inc/TBuffer3D.h
error: cannot open file '/opt/root_src/core/base/inc/TBuffer3D.h' for reading
Generating code for file core/base/inc/TBuffer3DTypes.h...
error: could not open file /opt/root_src/core/base/inc/TBuffer3DTypes.h
error: cannot open file '/opt/root_src/core/base/inc/TBuffer3DTypes.h' for reading
Generating code for file core/base/inc/TColor.h...
error: could not open file /opt/root_src/core/base/inc/TColor.h
error: cannot open file '/opt/root_src/core/base/inc/TColor.h' for reading
Generating code for file core/base/inc/TColorGradient.h...
error: could not open file /opt/root_src/core/base/inc/TColorGradient.h
error: cannot open file '/opt/root_src/core/base/inc/TColorGradient.h' for reading
Generating code for file core/base/inc/TDatime.h...
error: could not open file /opt/root_src/core/base/inc/TDatime.h
error: cannot open file '/opt/root_src/core/base/inc/TDatime.h' for reading
Generating code for file core/base/inc/TDirectory.h...
error: could not open file /opt/root_src/core/base/inc/TDirectory.h
error: cannot open file '/opt/root_src/core/base/inc/TDirectory.h' for reading
Generating code for file core/base/inc/TEnv.h...
error: could not open file /opt/root_src/core/base/inc/TEnv.h
error: cannot open file '/opt/root_src/core/base/inc/TEnv.h' for reading
Generating code for file core/base/inc/TException.h...
error: could not open file /opt/root_src/core/base/inc/TException.h
error: cannot open file '/opt/root_src/core/base/inc/TException.h' for reading
Generating code for file core/base/inc/TExec.h...
error: could not open file /opt/root_src/core/base/inc/TExec.h
error: cannot open file '/opt/root_src/core/base/inc/TExec.h' for reading
Generating code for file core/base/inc/TFileCollection.h...
error: could not open file /opt/root_src/core/base/inc/TFileCollection.h
error: cannot open file '/opt/root_src/core/base/inc/TFileCollection.h' for reading
Generating code for file core/base/inc/TFileInfo.h...
error: could not open file /opt/root_src/core/base/inc/TFileInfo.h
error: cannot open file '/opt/root_src/core/base/inc/TFileInfo.h' for reading
Generating code for file core/base/inc/TFolder.h...
error: could not open file /opt/root_src/core/base/inc/TFolder.h
error: cannot open file '/opt/root_src/core/base/inc/TFolder.h' for reading
Generating code for file core/base/inc/TInetAddress.h...
error: could not open file /opt/root_src/core/base/inc/TInetAddress.h
error: cannot open file '/opt/root_src/core/base/inc/TInetAddress.h' for reading
Generating code for file core/base/inc/TMacro.h...
error: could not open file /opt/root_src/core/base/inc/TMacro.h
error: cannot open file '/opt/root_src/core/base/inc/TMacro.h' for reading
Generating code for file core/base/inc/TMathBase.h...
error: could not open file /opt/root_src/core/base/inc/TMathBase.h
error: cannot open file '/opt/root_src/core/base/inc/TMathBase.h' for reading
Generating code for file core/base/inc/TMD5.h...
error: could not open file /opt/root_src/core/base/inc/TMD5.h
error: cannot open file '/opt/root_src/core/base/inc/TMD5.h' for reading
Generating code for file core/base/inc/TMemberInspector.h...
error: could not open file /opt/root_src/core/base/inc/TMemberInspector.h
error: cannot open file '/opt/root_src/core/base/inc/TMemberInspector.h' for reading
Generating code for file core/base/inc/TMessageHandler.h...
error: could not open file /opt/root_src/core/base/inc/TMessageHandler.h
error: cannot open file '/opt/root_src/core/base/inc/TMessageHandler.h' for reading
Generating code for file core/base/inc/TNamed.h...
error: could not open file /opt/root_src/core/base/inc/TNamed.h
error: cannot open file '/opt/root_src/core/base/inc/TNamed.h' for reading
Generating code for file core/base/inc/TNotifyLink.h...
error: could not open file /opt/root_src/core/base/inc/TNotifyLink.h
error: cannot open file '/opt/root_src/core/base/inc/TNotifyLink.h' for reading
Generating code for file core/base/inc/TObject.h...
error: could not open file /opt/root_src/core/base/inc/TObject.h
error: cannot open file '/opt/root_src/core/base/inc/TObject.h' for reading
Generating code for file core/base/inc/TObjString.h...
error: could not open file /opt/root_src/core/base/inc/TObjString.h
error: cannot open file '/opt/root_src/core/base/inc/TObjString.h' for reading
Generating code for file core/base/inc/TParameter.h...
error: could not open file /opt/root_src/core/base/inc/TParameter.h
error: cannot open file '/opt/root_src/core/base/inc/TParameter.h' for reading
Generating code for file core/base/inc/TPluginManager.h...
error: could not open file /opt/root_src/core/base/inc/TPluginManager.h
error: cannot open file '/opt/root_src/core/base/inc/TPluginManager.h' for reading
Generating code for file core/base/inc/TPoint.h...
error: could not open file /opt/root_src/core/base/inc/TPoint.h
error: cannot open file '/opt/root_src/core/base/inc/TPoint.h' for reading
Generating code for file core/base/inc/TPRegexp.h...
error: could not open file /opt/root_src/core/base/inc/TPRegexp.h
error: cannot open file '/opt/root_src/core/base/inc/TPRegexp.h' for reading
Generating code for file core/base/inc/TProcessID.h...
error: could not open file /opt/root_src/core/base/inc/TProcessID.h
error: cannot open file '/opt/root_src/core/base/inc/TProcessID.h' for reading
Generating code for file core/base/inc/TProcessUUID.h...
error: could not open file /opt/root_src/core/base/inc/TProcessUUID.h
error: cannot open file '/opt/root_src/core/base/inc/TProcessUUID.h' for reading
Generating code for file core/base/inc/TQClass.h...
error: could not open file /opt/root_src/core/base/inc/TQClass.h
error: cannot open file '/opt/root_src/core/base/inc/TQClass.h' for reading
Generating code for file core/base/inc/TQCommand.h...
error: could not open file /opt/root_src/core/base/inc/TQCommand.h
error: cannot open file '/opt/root_src/core/base/inc/TQCommand.h' for reading
Generating code for file core/base/inc/TQConnection.h...
error: could not open file /opt/root_src/core/base/inc/TQConnection.h
error: cannot open file '/opt/root_src/core/base/inc/TQConnection.h' for reading
Generating code for file core/base/inc/TQObject.h...
error: could not open file /opt/root_src/core/base/inc/TQObject.h
error: cannot open file '/opt/root_src/core/base/inc/TQObject.h' for reading
Generating code for file core/base/inc/TRedirectOutputGuard.h...
error: could not open file /opt/root_src/core/base/inc/TRedirectOutputGuard.h
error: cannot open file '/opt/root_src/core/base/inc/TRedirectOutputGuard.h' for reading
Generating code for file core/base/inc/TRef.h...
error: could not open file /opt/root_src/core/base/inc/TRef.h
error: cannot open file '/opt/root_src/core/base/inc/TRef.h' for reading
Generating code for file core/base/inc/TRefCnt.h...
error: could not open file /opt/root_src/core/base/inc/TRefCnt.h
error: cannot open file '/opt/root_src/core/base/inc/TRefCnt.h' for reading
Generating code for file core/base/inc/TRegexp.h...
error: could not open file /opt/root_src/core/base/inc/TRegexp.h
error: cannot open file '/opt/root_src/core/base/inc/TRegexp.h' for reading
Generating code for file core/base/inc/TRemoteObject.h...
error: could not open file /opt/root_src/core/base/inc/TRemoteObject.h
error: cannot open file '/opt/root_src/core/base/inc/TRemoteObject.h' for reading
Generating code for file core/base/inc/TROOT.h...
error: could not open file /opt/root_src/core/base/inc/TROOT.h
error: cannot open file '/opt/root_src/core/base/inc/TROOT.h' for reading
Generating code for file core/base/inc/TRootIOCtor.h...
error: could not open file /opt/root_src/core/base/inc/TRootIOCtor.h
error: cannot open file '/opt/root_src/core/base/inc/TRootIOCtor.h' for reading
Generating code for file core/base/inc/TStopwatch.h...
error: could not open file /opt/root_src/core/base/inc/TStopwatch.h
error: cannot open file '/opt/root_src/core/base/inc/TStopwatch.h' for reading
Generating code for file core/base/inc/TStorage.h...
error: could not open file /opt/root_src/core/base/inc/TStorage.h
error: cannot open file '/opt/root_src/core/base/inc/TStorage.h' for reading
Generating code for file core/base/inc/TString.h...
error: could not open file /opt/root_src/core/base/inc/TString.h
error: cannot open file '/opt/root_src/core/base/inc/TString.h' for reading
Generating code for file core/base/inc/TStringLong.h...
error: could not open file /opt/root_src/core/base/inc/TStringLong.h
error: cannot open file '/opt/root_src/core/base/inc/TStringLong.h' for reading
Generating code for file core/base/inc/TStyle.h...
error: could not open file /opt/root_src/core/base/inc/TStyle.h
error: cannot open file '/opt/root_src/core/base/inc/TStyle.h' for reading
Generating code for file core/base/inc/TSysEvtHandler.h...
error: could not open file /opt/root_src/core/base/inc/TSysEvtHandler.h
error: cannot open file '/opt/root_src/core/base/inc/TSysEvtHandler.h' for reading
Generating code for file core/base/inc/TSystem.h...
error: could not open file /opt/root_src/core/base/inc/TSystem.h
error: cannot open file '/opt/root_src/core/base/inc/TSystem.h' for reading
Generating code for file core/base/inc/TSystemDirectory.h...
error: could not open file /opt/root_src/core/base/inc/TSystemDirectory.h
error: cannot open file '/opt/root_src/core/base/inc/TSystemDirectory.h' for reading
Generating code for file core/base/inc/TSystemFile.h...
error: could not open file /opt/root_src/core/base/inc/TSystemFile.h
error: cannot open file '/opt/root_src/core/base/inc/TSystemFile.h' for reading
Generating code for file core/base/inc/TTask.h...

Thread 14 "doxygen" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff6a74700 (LWP 4696)]
0x0000555555a5d81a in std::vector<reg::PToken, std::allocator<reg::PToken> >::size (this=0x8) at /usr/include/c++/8/bits/stl_vector.h:806
806	      { return size_type(this->_M_impl._M_finish - this->_M_impl._M_start); }
```

Of course, all these files above exist and are readable. (It works well in single-thread).

If I build doxygen in Release Mode, I only see a couple of these messages before the crash, rather than so many.

**To Reproduce**
It can be reproduced by following the [build instructions](https://github.com/root-project/web/blob/main/for_developers/doxygen/index.md) and setting in the [Doxyfile](https://github.com/root-project/root/blob/master/documentation/doxygen/Doxyfile) NUM_PROC_THREADS to 7. 

**Version**
1.9.4 (f1e207abc9cec319793dea61ab990d866293bfed*)

**Stack trace**
[stacktrace.txt.zip](https://github.com/doxygen/doxygen/files/8135070/stacktrace.txt.zip)

**Additional context**
RAM memory is fine, no signs of memleaks.
I tried fixing our filter adding some read-write safeguards, but there is no difference in the final result. https://github.com/root-project/root/pull/9983

**Comment by ferdymercury** [comment]
Sorry, duplicate, GitHub was hanging

**Comment by doxygen** [comment]
@ferdymercury Still have to look in detail, but just looking at the output, I'd say the problem is that doxygen leaks open file handles until the system runs out (just a theory for now, to be confirmed).

**Comment by doxygen** [comment]
@ferdymercury Please check if the referenced commit fixes the problem. I have tested it on doxygen's own sources only so far, not yet on the root package.

**Comment by albert-github** [comment]
I saw that there was some things done with Fortran as well so I wrote a quick test procedure:
```
!> \file
!> Test for return type of Fortran


!> Normal subroutine
!> \param a the parameter
subroutine fie_ok(a)
  integer:: a
end subroutine

!> Return shown subroutine
!> \param a the parameter
!> \return should trigger a warning
subroutine fie_err(a)
  integer:: a
end subroutine
```
this gives in the old situation, correctly:
```
.../aa.f90:12: warning: documented empty return type of fie_err
```
but in the situation with the update the message is missing.

Example: [example.tar.gz](https://github.com/doxygen/doxygen/files/8148551/example.tar.gz)

**Comment by doxygen** [comment]
@albert-github I see the corresponding warning now only appears if `WARN_NO_PARAMDOC=YES`. 
Should be corrected with commit https://github.com/doxygen/doxygen/commit/379bd0eaad970358d83734be763e5b2d44b43e0c

Before that before, this warning was given as part of the detection routine, which was not the right place.

**Comment by albert-github** [comment]
When  I now have `WARN_NO_PARAMDOC=NO` I see the message regarding line 12 and I aslso see this message when `WARN_NO_PARAMDOC = YES` but in the later case I also see the warning:
```
.../aa.f90:6: warning: return type of member fie_ok is not documented
```
and this is not correct.

**Comment by doxygen** [comment]
@albert-github Indeed, made [this commit](https://github.com/doxygen/doxygen/commit/5461285db0c4035746238481842ff190c93ce2f1) to properly guard the warning. Let me know if I still missed anything.

**Comment by albert-github** [comment]
Checked with a bit more extensive example (but still small) and this looks OK.

**Comment by ferdymercury** [comment]
Thanks for the investigation!

I rebuilt, and I get this error printed out:
`Generating docs for fdoxygen: /opt/doxygen/src/dot.cpp:150: DotFilePatcher* DotManager::createFilePatcher(const QCString&): Assertion `rv.second' failed.`

It hangs there for a couple of minutes, and then it crashes as follows:
```
Makefile:49: recipe for target 'doxygen' failed
make: *** [doxygen] Aborted (core dumped)
22:59:48: The process "/usr/bin/make" exited with code 2.
```

**Comment by ferdymercury** [comment]
If I run with the debugger, I get:

```
doxygen: /opt/doxygen/src/dot.cpp:150: DotFilePatcher* DotManager::createFilePatcher(const QCString&): Assertion `rv.second' failed.
Generating docs for file graf3d/g3d/inc/THelix.h...
Generating docs for file graf3d/g3d/inc/THYPE.h...
Generating docs for file graf3d/g3d/inc/TMarker3DBox.h...
Generating docs for file graf3d/g3d/inc/TMaterial.h...

Thread 22 "doxygen" received signal SIGABRT, Aborted.
[Switching to Thread 0x7ffff426f700 (LWP 16662)]
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
51	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
```

The backtrace is attached here [bt.txt.zip](https://github.com/doxygen/doxygen/files/8149512/bt.txt.zip). Apparently there are 80+ threads being defined, even if I only specified NUM_PROC_THREADS = 7.

**Comment by doxygen** [comment]
@ferdymercury The dot administration was not properly protected indeed. Should be addressed with the latest commit. I still some threading issues, so please try and keep reporting any issues you see. Note that the number of threads used by dot is determined by `DOT_NUM_THREADS`, so a different setting. If the administration is corrupted, the number of thread could be more than it should.

**Comment by ferdymercury** [comment]
Thanks so much for the fix. Now I run it once and it did not crash. :)

However, when running it a second time, it was hanging here for minutes without printing anything else:
```
...
Generating docs for compound TEnumConstant...
Generating docs for compound TEnv...
^C^C
```

I pressed CTRL+C many times, and after some minutes of waiting, it said:
`make: *** [doxygen] Segmentation fault (core dumped)`

The third and fourth time I run it, it was perfectly fine.

Best regards.

**Comment by albert-github** [comment]
That second run is a bit disturbing, good that the 3rd and 4th run did run.
- is the sequence repeatable? i.e. when starting from scratch and do the same as before i.e. running 4 times?
- did you remove the result directories between the runs ?

**Comment by ferdymercury** [comment]
No, I did not remove the results directory. I will try again.
I am also wondering: maybe there is also something 'thread-unsafe' in our custom filter executable. Or is the multi-threading only done for reading the input but not for the filtering ?
Thanks for the help.

**Comment by albert-github** [comment]
I don't think your filter will be "thread-unsafe" as the filters are invoked as sub processes, so they shouldn't know anything about threads.

**Comment by ferdymercury** [comment]
Yes, but the filter generates sometimes a temporary output file called ImageSizes.dat, which has the same name for every process, and then opens it later from another function. If two processes try to write at the same time, it might get a read/write a wrong number from ImageSizes.dat. But maybe it will not lead to a crash.

Here a stacktrace that Ubuntu generated from yesterday's crash (I believe).
![image](https://user-images.githubusercontent.com/10653970/156145867-245467e4-8074-42c7-9038-e6f5e0c9a080.png)

![image](https://user-images.githubusercontent.com/10653970/156146047-5daabb3f-7bb1-46de-a842-0ef969eb86f5.png)

![image](https://user-images.githubusercontent.com/10653970/156146219-b0ae55e0-df79-4fdd-8974-145ed9df524f.png)

**Comment by albert-github** [comment]
Regarding:
> Yes, but the filter generates sometimes a temporary output file called ImageSizes.dat, which has the same name for every process, and then opens it later from another function. If two processes try to write at the same time, it might get a read/write a wrong number from ImageSizes.dat. But maybe it will not lead to a crash.


This can lead to very strange results in the output and even crashes. These kind of problems are very hard (nearly impossible) to track down but might lead to the crash you saw (due to some side effects during the earlier processing).
Can't you use here a filename that includes e.g. the process number (this will be the process number of the filter executable) so that it will be unique or use the `tmpfile` function.

**Comment by ferdymercury** [comment]
Yes, I will change that. Thanks for the hint.

**Comment by doxygen** [comment]
Some more data races caught while doing more extensive testing.

**Comment by ferdymercury** [comment]
Thanks again!
I have fixed also the unsafe parts on my side, and now it has run 4+ times in a row without problems. :)

**Comment by albert-github** [comment]
Looks like that the issues mentioned in this issue have been fixed, setting the issue to "fixed but not released".

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.4.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
